File: src/leetcode/CharacterReplacement.jl
function characterReplacement(s::String, k::Int)::Int
    ## initialize the variables
    char_count = Dict{Char, Int}() # A Dict to store the count of each character in the window
    L = 1 # Left pointer of the window
    max_count = 0 # The count of the most frequent character in the window
    max_length = 0 # The length of the longest substring with the same character

    ## Loop through the string
    for R in eachindex(s)
        char = s[R]
        if haskey(char_count, char)
            char_count[char] += 1
        else
            char_count[char] = 1
        end
        # update the max count character in the window
        max_count = max(max_count, char_count[char])

        current_length = R - L + 1

        if current_length - max_count > k
            char_count[s[L]] -=1
            L += 1 # shrink the substring by moving the left pointer to right
        else
            max_length = max(max_length, current_length)
        end

    end
    return max_length
end


if abspath(PROGRAM_FILE) == @__FILE__
    println(characterReplacement("ABAB", 2))  # Expected output: 4
    println(characterReplacement("AABABBA", 1)) # Expected output: 4
end

File: src/leetcode/python/sliding_window/character_replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        # Initialize a dict to store the character counts
        char_count = {}

        # initialize a left pointer
        l = 0
        max_f = 0
        res = 0
        # initialize the right pointer at the beginning as well
        # loop through the entire string
        for r in range(len(s)):
            # increment the count as the right pointer moves
            char_count[s[r]] = 1 + char_count.get(s[r], 0)
            max_f = max(max_f, char_count[s[r]])

            if (r - l + 1) - max_f > k:
                l += 1
                char_count[s[l]] -= 1

            res = max(res, r - l + 1)

        return res


if __name__ == "__main__":
    pass
